[ONNX] Refactor `TestONNXRuntime` with parameterized class

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #79641

* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

BowenBao committed:
[ONNX] Refactor `TestONNXRuntime` with parameterized class

* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

justinchuby commented on test/onnx/test_models_onnxruntime.py:
```diff
@@ -4,16 +4,19 @@
-from test_pytorch_common import skipScriptTest_New as skipScriptTest
-from test_pytorch_onnx_onnxruntime import run_model_test
+from test_pytorch_common import (
+    TestCase,
```
It's good time we import from common_utils instead:

```python
from torch.testing._internal import common_utils
```

and do 

```
class TestVerification(common_utils.TestCase):
```
TestCase was transitively imported in test_pytorch_common and we should avoid that. related #79543

BowenBao replied:
let's leave this to the other PR then. The diff is from unrelated changes on `run_model_test`.

justinchuby commented on test/onnx/test_models_onnxruntime.py:
```diff
@@ -36,11 +39,15 @@ def exportTest(self, model, inputs, rtol=1e-2, atol=1e-7, opset_versions=None):
         self.opset_version = opset_version
         self.onnx_shape_inference = True
-        run_model_test(self, model, input_args=inputs, rtol=rtol, atol=atol)
+        test_onnx_common._run_model_test(
```
_run_model_test can be public, since: it is in test, and it is used in other modules.

justinchuby commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -12351,139 +12277,33 @@ def make_test(
+            ("elman", "tanh"),
+            ("lstm", None),
+            ("gru", None),
+        ],
```
Do we know if the order of the decorators matter? Do we need the parameterization to be the last applied (placed at the top)?

BowenBao replied:
it's kwargs, and it's written that names appear in top to bottom order. see `compose_parametrize_fns`.

justinchuby replied:
I meant the order of `common_utils.parametrize` and the skips

BowenBao replied:
I was referring to the same thing.

justinchuby replied:
Perfect thanks!

justinchuby commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -12351,139 +12277,33 @@ def make_test(
-
-TestONNXRuntime_opset7 = MakeTestCase(7)
-
-TestONNXRuntime_opset8 = MakeTestCase(8)
```
Deleting code. This feels good

justinchuby commented on test/onnx/test_onnx_common.py:
```diff
@@ -38,6 +39,11 @@ def _run_model_test(test_suite: _TestONNXRuntime, *args, **kwargs):
     return verification.verify(*args, **kwargs)
 
 
+def parameterize_class_name(cls: Type, idx: int, input_dicts: Mapping[Any, Any]):
```
I would add a docstring to this function to make it clear how anyone can use it. especially since it is shared

justinchuby commented on test/onnx/test_onnx_common.py:
```diff
@@ -38,6 +39,11 @@ def _run_model_test(test_suite: _TestONNXRuntime, *args, **kwargs):
 
 
+def parameterize_class_name(cls: Type, idx: int, input_dicts: Mapping[Any, Any]):
+    suffix = "_".join(f"{k}_{v}" for k, v in input_dicts.items())
```
Do we need to use `parameterized.to_safe_name`?

BowenBao replied:
Didn't include it for now for simplicity. The method itself is unnecessarily a classmethod unfortunately. We could utilize it or do something similar, if someone needs to add parameters with nonconforming characters in the future.

justinchuby reviewed (commented):
Love this PR!

justinchuby commented:
`ImportError: cannot import name 'run_model_test' from 'test_pytorch_onnx_onnxruntime'`

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -49,31 +49,14 @@
 
 import torch
 import torch.nn.functional as F
+import torch.onnx._constants as _constants
```
weird, but anyway the import won't be needed.

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

justinchuby commented on test/onnx/test_custom_ops.py:
```diff
@@ -2,14 +2,14 @@
 import numpy as np
 import onnx
+from test_onnx_common import run_model_test
+from test_pytorch_common import TestCase, run_tests
```
import from torch testing internals

BowenBao replied:
same to above

justinchuby replied:
Sounds good

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

justinchuby reviewed (approved):
LGTM🚀

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Refactor `TestONNXRuntime` with parameterized class"


* Refactor `TestONNXRuntime` with parameterized class, making it obvious whether
  the unittest is testing on TorchScript or trace based export.
* Refactor rnn_tests with parametrized test cases.

Fixes #75630

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g
